File: test/corpus/propagation/impl_into_option_propagation.exs
defmodule RustQ.Corpus.Propagation.ImplIntoOptionPropagation do
  @moduledoc "Propagate fallible values passed to impl Into<Option<T>> arguments."

  use RustQ.Meta, rust_sources: ["test/fixtures/into_option_method.rs"]

  alias RustQ.Type, as: R

  @spec decode_filter(term()) :: R.nif_result(R.path(:ImageFilter))
  defrust decode_filter(_term) do
    {:ok, ImageFilter.default()}
  end

  @spec run(R.mut_ref(R.path(:Paint)), term()) :: R.nif_result(R.unit())
  defrust run(paint, term) do
    paint.set_image_filter(decode_filter(term))
    :ok
  end
end

File: test/corpus/propagation/impl_into_option_propagation.rs
fn decode_filter<'a>(_term: Term<'a>) -> NifResult<ImageFilter> {
    Ok(ImageFilter::default())
}

fn run<'a>(paint: &mut Paint, term: Term<'a>) -> NifResult<()> {
    paint.set_image_filter(decode_filter(term)?);
    Ok(())
}
